truffle/src/main/java/org/jruby/truffle/nodes/ext/psych/PsychEmitterNodes.java

```diff
@@ -0,0 +1,493 @@
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.object.DynamicObject;
+import com.oracle.truffle.api.source.SourceSection;
+import org.jcodings.Encoding;
+import org.jruby.runtime.Visibility;
+import org.jruby.truffle.nodes.core.CoreClass;
+import org.jruby.truffle.nodes.core.CoreMethod;
+import org.jruby.truffle.nodes.core.CoreMethodArrayArgumentsNode;
+import org.jruby.truffle.nodes.objects.AllocateObjectNode;
+import org.jruby.truffle.nodes.objects.AllocateObjectNodeGen;
+import org.jruby.truffle.runtime.NotProvided;
+import org.jruby.truffle.runtime.RubyContext;
+import org.jruby.truffle.runtime.adapaters.OutputStreamAdapter;
+import org.jruby.truffle.runtime.core.ArrayOperations;
+import org.jruby.truffle.runtime.layouts.Layouts;
+import org.yaml.snakeyaml.DumperOptions;
+import org.yaml.snakeyaml.emitter.Emitter;
+import org.yaml.snakeyaml.emitter.EmitterException;
+import org.yaml.snakeyaml.error.Mark;
+import org.yaml.snakeyaml.events.*;
+
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.nio.charset.Charset;
+import java.util.HashMap;
+import java.util.Map;
+
+@CoreClass(name = "Psych::Emitter")
+public abstract class PsychEmitterNodes {
+
+    @CoreMethod(names = "allocate", constructor = true)
+    public abstract static class AllocateNode extends CoreMethodArrayArgumentsNode {
+
+        @Child private AllocateObjectNode allocateNode;
+
+        public AllocateNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+            allocateNode = AllocateObjectNodeGen.create(context, sourceSection, null, null);
+        }
+
+        @Specialization
+        public DynamicObject allocate(DynamicObject rubyClass) {
+            return allocateNode.allocate(rubyClass, null, null, null);
+        }
+
+    }
+
+    @CoreMethod(names = "initialize", visibility = Visibility.PRIVATE, required = 1, optional = 1)
+    public abstract static class InitializeNode extends CoreMethodArrayArgumentsNode {
+
+        public InitializeNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+        }
+
+        @Specialization
+        public DynamicObject initialize(DynamicObject emitter, DynamicObject io, NotProvided optionsSet) {
+            final DumperOptions options = new DumperOptions();
+            options.setIndent(2);
+            Layouts.PSYCH_EMITTER.setOptions(emitter, options);
+
+            Layouts.PSYCH_EMITTER.setIo(emitter, io);
+
+            return nil();
+        }
+
+        @CompilerDirectives.TruffleBoundary
+        @Specialization
+        public DynamicObject initialize(DynamicObject emitter, DynamicObject io, DynamicObject optionsSet) {
+            final DumperOptions options = new DumperOptions();
+            options.setWidth((int) ruby("options_set.line_width", "options_set", optionsSet));
```

Comment:
Maybe we should have a `send` helper in this case: `send(optionsSet, "line_width")`.